agent/agent/process/classification_and_content.py

```diff
@@ -0,0 +1,198 @@
+from arxiv.submission.domain.annotation import Feature
+from arxiv.submission.domain.agent import Agent, User
+from arxiv.submission.domain.process import ProcessStatus
+from arxiv.submission.services import Classifier, PlainTextService
+from arxiv.submission.services.plaintext import ExtractionFailed
+
+from .base import Process, step, Retry, Recoverable
+from ..domain import Trigger
+
+
+class PlainTextExtraction(Process):
+    """Extract plain text from a compiled PDF."""
+
+    def source_id(self, trigger: Trigger) -> int:
+        """Get the source ID for the submission content."""
+        try:
+            return trigger.after.source_content.identifier
+        except AttributeError as exc:
+            self.fail(exc, 'No source content identifier on post-event state')
+
+    def handle_plaintext_exception(self, exc: Exception) -> None:
+        """Handle exceptions raised when calling the plain text service."""
+        exc_type = type(exc)
+
+        if exc_type in (exceptions.BadResponse, exceptions.ConnectionFailed):
+            raise Recoverable('Encountered %s; try again' % exc) from exc
+        elif exc_type is ExtractionFailed:
+            self.fail(exc, 'Extraction service failed to extract text')
+        elif exc_type is exceptions.RequestFailed:
+            if exc.status_code >= 500:
+                msg = 'Extraction service choked: %i' % exc.status_code
+                raise Recoverable(msg) from exc
+            self.fail(exc, 'Unrecoverable exception: %i' % exc.status_code)
+        self.fail(exc, 'Unhandled exception')
+
+    @step(max_retries=None)
+    def start_extraction(self, previous: Optional, trigger: Trigger,
+                         emit: Callable) -> None:
+        """Request extraction by the plain text service."""
+        try:
+            PlainTextService.request_extraction(self.source_id(trigger))
+        except Exception as exc:
+            self.handle_plaintext_exception(exc)
+
+    @step(max_retries=None, delay=1, backoff=1, jitter=(0, 1))
+    def poll_extraction(self, previous: Optional, trigger: Trigger,
+                        emit: Callable) -> None:
+        """Poll the plain text service until extraction is complete."""
+        source_id = self.source_id(trigger)
+        try:
+            complete = PlainTextService.extraction_is_complete(source_id)
+        except Exception as exc:
+            self.handle_plaintext_exception(exc)
+        if not complete:
+            raise Retry('Not complete; try again')
+
+    @step(max_retries=None)
+    def retrieve_content(self, previous: Optional, trigger: Trigger,
+                         emit: Callable) -> bytes:
+        """Retrieve the extracted plain text."""
+        source_id = self.source_id(trigger)
+        try:
+            return PlainTextService.retrieve_content(source_id)
+        except Exception as exc:
+            self.handle_plaintext_exception(exc)
+
+
+class RunAutoclassifier(PlainTextExtraction):
```

Comment:
Is this code functional with current autoclassifier, or should we raise a `NotImplementedError` if we try to run the autoclassifier? Assuming it will be migrated to an event architecture later.

Reply:
It is functional.